### DISCUSSION TITLE
it is not possible to get current objective bounds between solutions

### DISCUSSION THREAD

**Discussion by gregy4** [google/or-tools#4202] [Feature requests] [answered]
**What language and solver does this apply to?**
Python, Java / CP-SAT

**Describe the problem you are trying to solve.**
It is not possible to get current objective bounds between solutions. It is only shown in a log.

**Describe the solution you'd like**
New method in solution callback. Executed when new bound of objective function is found.
  or
Update value of best objective bound so it can be read externally. So far it seems it is updated only when a solution is found.

Add relative gap in a log for #bound event, especially for large domains it can be more helpful then numbers:
      #4       0.07s best:3.837456e+10 next:[2.18861401e+10,3.83745599e+10] graph_var_lns (d=0.50 s=39 t=0.10 p=0.00 stall=0 h=auto_l0) [hint]

**Describe alternatives you've considered**
I tried to use external timer but best objective bound isn't updated between solutions.

**Comment by gregy4** [comment]
Why was the issue #4201 moved to discussions ?

It is quite easy to be informed about new bounds of objective function.
The method SharedResponseManager::UpdateInnerObjectiveBounds should call a callback when new bound is found. The value is already written in new_lb variable for logging->LoggingIsEnabled && change.
Experimentally I reused CpSolverResponse and updated only best_objective_bound. It works for me.

      CpSolverResponse result =
      solutions_.NumSolutions() == 0
          ? GetResponseInternal({}, "")
          : GetResponseInternal(solutions_.GetSolution(0).variable_values,
                                solutions_.GetSolution(0).info);

      result.set_best_objective_bound(new_lb);

      for (const auto& pair : callbacks_) {
        pair.second(result);
      }

Adding of relative gap in a log is especially important for setting of relative_gap_limit. Otherwise a user needs to parse log file or manually calculate gap on every bound event to know when a solver found that a solution is in given distance from optimal solution. The change is trivial, again in the method UpdateInnerObjectiveBounds.

**Comment by lperron** [comment]
I move all feature requests to discussions/feature requests.

I still use the issue template as I have not yet looked at discussions templates, not do I know if I can create a discussion from an issue template.

**Comment by lperron** [comment]
and to answer your question, it is not a zero cost effort.

1) I need to wrap it in all languages
2) There is a throttling mechanism in place as I have models which generates millions of improving lower bounds
3) logging slows down the solver a lot. You need to postprocess all incumbent solutions, which is very costly

**Reply by gregy4** [reply]
I understand that.
Maybe light version with update of SolutionCallback::BestObjectiveBound() would be enough so current value can be checked externally. In this case 1. is not needed. Update can be explicitly enabled by solver parameter to not affect models mentioned in 2.

**Comment by lperron** [comment]
for the gap printing, there is a solution log callback code where you can append the mip gap for each solution.

**Comment by lperron** [comment]
not trivial.
SolutionCallback is a non C++ stub.
In C++, there are no wrapping classes, just lambdas.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment, marked as answer]
I just pushed the code.
